### PR TITLE
Assert that the file isnt open in lfs_file_opencfg

### LINKED ISSUE #429: infinite loop during file close
if lfs_file_open is called two times (with identical arguments) it will return ok to both calls. That will create self referencing linked list (lfs->mlist->next will look to itself). During file closing (in function lfs_dir_commit) linked list will create infinite loop and never return.

I'm aware that described usage is not valid but maybe you could add assertion or return error result on second call. I discovered that behavior when I've tested lfs_file_opencfg bug forgot to remove previous lfs_file_open.

### PR DESCRIPTION
**Pull request by NoahGorny**
Closes #429

Also fixed a magic number in `opencfg`

### CODE REVIEW

**Comment by e107steved** [comment]
I think I'd prefer the operation failed by returning an error, rather than doing an assert

**Comment by NoahGorny** [comment]
> I think I'd prefer the operation failed by returning an error, rather than doing an assert

I think that checking every time for something that is really a user error is not needed, but I will accept this if more people prefer checks. If your program hangs in littlefs I expect that you will compile littlefs in debug and then you will see the problem.

In a different matter, I really need to find a way to check if the struct was initialized, because right now I access uninitialized memory :(

**Comment by judgeh** [comment]
I definitely prefer the return code. I think asserts are heavily overused in lfs. there are many cases where my application would benefit from handling return codes rather than the asserting.

File: lfs.c [hunk 1]
```diff
@@ -2387,9 +2398,10 @@ int lfs_file_opencfg(lfs_t *lfs, lfs_file_t *file,
                  ".buffer=%p, .attrs=%p, .attr_count=%"PRIu32"})",
             (void*)lfs, (void*)file, path, flags,
             (void*)cfg, cfg->buffer, (void*)cfg->attrs, cfg->attr_count);
+    LFS_ASSERT(is_node_not_in_list(lfs->mlist, (struct lfs_mlist*)file));
```
Comment (diederikvdv) [inline lfs.c:2401]: This check can also be added to the lfs_dir_open function:
LFS_ASSERT(is_node_not_in_list(lfs->mlist, (struct lfs_mlist*)dir));

This requires the static function to be declared earlier. Other static functions are declared at the top of the file

Comment (NoahGorny) [reply lfs.c:2403]: added, thanks for the feedback!

**Comment by NoahGorny** [comment]
Guys, I am not sure why you should return error on such cases. Iterating over all open files in each open call seems costly, so I think we can stick to an assert. You can open the same file twice, just not with the same struct...

**Comment by judgeh** [comment]
> You can open the same file twice, just not with the same struct...

Oh in our testing, we were unable to open a file twice with different structures.
we
1. opened a file with hdl1
2. wrote data using hdl1
3. synced hdl1 (but not close) 
4. opened same file with hdl2
5. read hd2 (this always reported zero bytes) 

Do you expect, in this scenario, the hdl2 read to return the bytes written by hdl1 ? Maybe our test case was buggy, but we concluded that we could not read a file that was already open. If this is not the case, then thats great.

**Comment by judgeh** [comment]
Also, isnt it still checking every open file in the assert? How is asserting any different than returning in terms on complexity ?

**Comment by NoahGorny** [comment]
I will check the test case you mentioned, thanks!
I currently only check that the struct is not reused, in order to test that the file is not already opened, I will need to add a new assert.

And assert is better because it is not complied in, unless on debug

**Comment by geky** [comment]
Hi @NoahGorny, thanks for creating a PR! Sorry to leave this hanging for so long. This is a good addition.

I wonder if it would also be good to replace LFS_F_OPENED (https://github.com/littlefs-project/littlefs/pull/237) so we have the same check for open/close files that doesn't get tripped up by uninitialized memory. Thoughts?

---

My opinion is this check should assert. For a code size reason, not a usability reason. Asserts can be removed during compilation, return codes can not. And for microcontroller-scale programs, you really don't want to pay for code that should never run.

Also the current behavior for these types of user checks is to assert, if we want to change this behavior it should be done via a different issue or PR.

File: lfs.c [hunk 2]
```diff
@@ -2387,9 +2400,10 @@ int lfs_file_opencfg(lfs_t *lfs, lfs_file_t *file,
...
 
     // deorphan if we haven't yet, needed at most once after poweron
-    if ((flags & 3) != LFS_O_RDONLY) {
+    if ((flags & LFS_O_RDWR) != LFS_O_RDONLY) {
```
Comment (geky) [inline lfs.c:2406]: 👍

File: lfs.c [hunk 3]
```diff
@@ -411,6 +411,17 @@ static inline void lfs_superblock_tole32(lfs_superblock_t *superblock) {
     superblock->attr_max    = lfs_tole32(superblock->attr_max);
 }
 
+static inline bool is_node_not_in_list(struct lfs_mlist *head,
+    struct lfs_mlist *node) {
```
Comment (geky) [inline lfs.c:415]: style nit: Two indents when inside parens (I realize I don't have this documented anywhere...)

File: lfs.c [hunk 4]
```diff
@@ -411,6 +411,17 @@ static inline void lfs_superblock_tole32(lfs_superblock_t *superblock) {
     superblock->attr_max    = lfs_tole32(superblock->attr_max);
 }
 
+static inline bool is_node_not_in_list(struct lfs_mlist *head,
```
Comment (geky) [inline lfs.c:414]: nit: Prefix with `lfs_`. Thoughts on `lfs_mlist_isopen` for consistency with the other functions?

**Comment by diederikvdv** [comment]
> I wonder if it would also be good to replace LFS_F_OPENED (#237) so we have the same check for open/close files that doesn't get tripped up by uninitialized memory. Thoughts?

If LFS_F_OPENED (#237) is replaced, it might also be a good addition to add the isOpened API mentioned in #442. But then based on lfs_is_node_not_in_list instead of LFS_F_OPENED

File: lfs.c (hunk 2 above)
Comment (NoahGorny) [reply lfs.c:2410]: I changed in all places now :smile:

File: lfs.c (hunk 3 above)
Comment (NoahGorny) [reply lfs.c]: done

File: lfs.c (hunk 4 above)
Comment (NoahGorny) [reply lfs.c]: done, changed to `lfs_mlist_isused`

**Comment by NoahGorny** [comment]
> Hi @NoahGorny, thanks for creating a PR! Sorry to leave this hanging for so long. This is a good addition.
> 
> I wonder if it would also be good to replace LFS_F_OPENED (#237) so we have the same check for open/close files that doesn't get tripped up by uninitialized memory. Thoughts?

I agree, and added commits that deprecate LFS_F_OPENED. It required me to sometimes add files to the mlist when we did not do so before, so you might wanna take a look at 0250015 and 0ca3012 to tell me whether my solution is okay or not.  
In any case, I did not add support for asserting of file that was opened twice using a different `file` struct, so we might want to take a look at it. Reusing an existing struct will now raise assert errors :smile:  
By the way, are there any ways to "catch' asserts in the tests? I want to deliberately see the asserts trigger in the tests.

Thanks alot for your feedback btw @geky

**Comment by NoahGorny** [comment]
> > I wonder if it would also be good to replace LFS_F_OPENED (#237) so we have the same check for open/close files that doesn't get tripped up by uninitialized memory. Thoughts?
> 
> If LFS_F_OPENED (#237) is replaced, it might also be a good addition to add the isOpened API mentioned in #442. But then based on lfs_is_node_not_in_list instead of LFS_F_OPENED

I think @geky should voice his opinion here. I think that currently with the assert not fully catching all cases (opened with different struct) it should not be public API yet... maybe after we catch/handle all such cases

**Comment by NoahGorny** [comment]
> > You can open the same file twice, just not with the same struct...
> 
> Oh in our testing, we were unable to open and read a file twice with different structures.
> we
> 
>     1. opened a file with hdl1
> 
>     2. wrote data using hdl1
> 
>     3. synced hdl1 (but not close)
> 
>     4. opened same file with hdl2
> 
>     5. read hd2 (this always reported zero bytes)
> 
> 
> Do you expect, in this scenario, the hdl2 read to return the bytes written by hdl1 ? Maybe our test case was buggy, but we concluded that we could not read a file that was already open. If this is not the case, then thats great.

@judgeh It indeed fails with my tests, I created a test (on a separate PR) that fails on this (ad995ceb9bf72f5c87690e2c63f70ed585fabb29)  
I am not sure what is the correct approach to this, but I think @geky can have some insights :smile:

**Comment by geky** [comment]
> Oh in our testing, we were unable to open and read a file twice with different structures.
> > @judgeh It indeed fails with my tests, I created a test (on a separate PR) that fails on this (ad995ce)

Ah, this looks like the same issue being discussed in https://github.com/littlefs-project/littlefs/issues/483. Thumbing up https://github.com/littlefs-project/littlefs/issues/483 would help know who's ran into this. Also thanks for the test case!

TLDR: Currently, littlefs is transactional in the sense that each open file handle gets its own snapshot. Each file handle can be read/written independently (and independently from disk). Something that came up in https://github.com/littlefs-project/littlefs/issues/483, we could go through all open files and throw out other changes on sync/close, it would just take more code size. More info in https://github.com/littlefs-project/littlefs/issues/483.

File: lfs.c [hunk 5]
```diff
@@ -411,6 +411,22 @@ static inline void lfs_superblock_tole32(lfs_superblock_t *superblock) {
...
+    for (struct lfs_mlist **p = &head; *p; p = &(*p)->next) {
+        if (*p == (struct lfs_mlist*)node) {
+            return true;
+        }
+    }
+
+    return false;
+}
+
+static inline void lfs_mlist_add(lfs_t *lfs, struct lfs_mlist *mlist) {
```
Comment (geky) [inline lfs.c:425]: If we add this, `lfs_mlist_remove` would be more valuable:
[https://github.com/littlefs-project/littlefs/blob/4c9146ea539f72749d6cc3ea076372a81b12cb11/lfs.c#L2553-L2559 ](https://github.com/littlefs-project/littlefs/blob/4c9146ea539f72749d6cc3ea076372a81b12cb11/lfs.c#L2553-L2559)

Also a nit: I think `lfs_mlist_append` captures the data structure operation better.

File: lfs.h [hunk 6]
```diff
@@ -136,7 +136,7 @@ enum lfs_open_flags {
     LFS_F_READING = 0x040000, // File has been read since last flush
     LFS_F_ERRED   = 0x080000, // An error occured during write
     LFS_F_INLINE  = 0x100000, // Currently inlined in directory entry
-    LFS_F_OPENED  = 0x200000, // File has been opened
+    LFS_F_OPENED  = 0x200000, // DEPRECATED: File has been opened
```
Comment (geky) [inline lfs.h:139]: Since these are marked as "internally used flags", I think it's safe to remove without deprecation.

File: lfs.c [hunk 7]
```diff
@@ -411,6 +411,22 @@ static inline void lfs_superblock_tole32(lfs_superblock_t *superblock) {
     superblock->attr_max    = lfs_tole32(superblock->attr_max);
 }
 
+static inline bool lfs_mlist_isused(struct lfs_mlist *head,
```
Comment (geky) [inline lfs.c:414]: nit: I think this would be better as `lfs_mlist_isopen`, it is directly related to `lfs_file_open` and `lfs_dir_open`.

File: lfs.c (hunk 2 above)
Comment (geky) [reply lfs.c:2410]: Ah I misread this. Well, I don't really have an opinion on this change. It is a curious mask.

File: lfs.c [hunk 8]
```diff
@@ -2400,23 +2418,22 @@ int lfs_file_opencfg(lfs_t *lfs, lfs_file_t *file,
...
-    file->flags = flags | LFS_F_OPENED;
     file->pos = 0;
     file->off = 0;
+    file->flags = flags;
     file->cache.buffer = NULL;
 
+    // get id, add to list of mdirs to catch update changes
+    file->type = LFS_TYPE_REG;
+    lfs_mlist_add(lfs, (struct lfs_mlist *)file);
+
```
Comment (geky) [inline lfs.c:2429]: I wouldn't move this either, the `lfs_mdir_t` is not initialized until the `lfs_dir_find`, and I'm not sure having an uninitialized mdir on in mlist is safe.

Unfortunately the mlist is a bit sensitive because all wear-leveling/allocation logic depend on it. It would be better if we prefer changing the LFS_F_OPENED behavior slightly to avoid changing the mlist logic, since LFS_F_OPENED asserts are just protection against already invalid API usage.

**Comment by geky** [comment]
> it might also be a good addition to add the isOpened API mentioned in #442

I'm still not convinced on https://github.com/littlefs-project/littlefs/issues/442. The behavior of an `lfs_file_isopened` could be accomplished with an external flag without requiring an additional API into the filesystem. It could also be added later if there is interest.

**Comment by geky** [comment]
> By the way, are there any ways to "catch' asserts in the tests?

Almost missed this, unfortunately no, the test script currently only treats asserts as failures. The test script should also be documented better...

File: lfs.c (hunk 2 above)
Comment (NoahGorny) [reply lfs.c:2410]: I think this is better than a magic num

File: lfs.c (hunk 5 above)
Comment (NoahGorny) [reply lfs.c]: done both!

File: lfs.h (hunk 6 above)
Comment (NoahGorny) [reply lfs.h]: removed the change entirely

File: lfs.c (hunk 7 above)
Comment (NoahGorny) [reply lfs.c:414]: its okay, changed it

File: lfs.c (hunk 8 above)
Comment (NoahGorny) [reply lfs.c]: its honestly better to leave alone the LFS_F_OPENED flag honestly. It causes those problems with the assert and the `isused` check can be utilized only when need (in `file_opencfg` and `dir_open`) and the rest of the API can stay with the flag

File: lfs.c (hunk 7 above)
Comment (geky) [reply lfs.c:414]: It looks like this is still `lfs_mlist_isused`, or am I missing a commit?

File: lfs.c (hunk 2 above)
Comment (geky) [reply lfs.c:2414]: Can't argue with that

File: lfs.c (hunk 7 above)
Comment (NoahGorny) [reply lfs.c]: sorry, updated it now :smile:

**Review by geky — APPROVED** [review]
This looks great, thanks for this! Will merge into the next release.

**Comment by geky** [comment]
I didn't notice until after merging that this reverted the `LFS_F_OPENED` flag, so I went ahead and re-added that change to #495. But can re-revert if needed.

I think I ran into some of the issues you did with the change, but one thing that helped was the new separate API-layer, I could move the check there and it solved some of complex internal situations, as well as reduced the calls to `lfs_mlist_isopen` to the API layer.
